File: BOJ/25_최단경로/10217_KCM_Travel.cpp
#include <vector>
#include <queue>
#include <iostream>
#include <algorithm>
using namespace std;

/*
    다익스트라로 풀수 있다!!! -> dist 배열을 잘 구현해줘야함.
    43-47 : 적은 예산으로 적은 시간이면 이후에 예산을 더썼을 때도 돌 필요 없이 만들어줍시다.
*/

#define MAXN 100
#define MAXM 100000
#define INF 987654321
vector<pair<int,pair<int,int> > > adjs[MAXN+1];
int dist[MAXN+1][MAXM+1];
int T,N,M,K,u,v,c,d;

int dji() {
    for(int i=0; i<=N; i++) {
        for(int j=0; j<=M; j++){
            dist[i][j] = INF;
        }
    }
    dist[1][0] = 0;
    priority_queue<pair<int, pair<int,int> > > pq; // time, cost, here
    pq.push(make_pair(0,make_pair(0,1)));

    while(!pq.empty()) {
        int time = -pq.top().first;
        int cost = pq.top().second.first;
        int here = pq.top().second.second;
        pq.pop();
        if(dist[here][cost] < time) continue;

        for(int i=0; i<adjs[here].size(); i++) {
            int next = adjs[here][i].first;
            int nextCost = cost + adjs[here][i].second.first;
            int nextTime = time + adjs[here][i].second.second;
            
            if (nextCost > M) continue;
            if(dist[next][nextCost] > nextTime) {
                for(int j=nextCost; j<=M; j++){
                    if(dist[next][j] > nextTime){
                        dist[next][j] = nextTime;
                    }
                }
                pq.push(make_pair(-nextTime,make_pair(nextCost,next)));
            } 
        }
    }
    return dist[N][M];
}

int main() {
    
    scanf("%d",&T);
    while(T--) {
        scanf("%d %d %d",&N,&M,&K);
        for(int i=0; i<=MAXN; i++) adjs[i].clear();
        for(int i=0; i<K; i++) {
            scanf("%d %d %d %d",&u,&v,&c,&d);
            adjs[u].push_back(make_pair(v, make_pair(c,d)));
        }
        int ans = dji();
        if(ans == INF) printf("Poor KCM\n");
        else printf("%d\n",ans);
    }


    return 0;
}

File: BOJ/25_최단경로/10217_KCM_Travel.py
import sys
import heapq
read = sys.stdin.readline
INF = int(1e9)
'''
냅색 문제 처럼 풀었는데 최소힙을 사용해서 풀수 있지않을까.ㅜ 
python 시간초과 , pypy 통과
dist 배열에서 아이디어 얻어서 최소힙 사용해서 품 -> 시간초과. 
같은 알고리즘으로 C++ 은 통과
'''
def dji():
    dist = [[INF] * (M+1) for _ in range(N+1)]
    dist[1][0] = 0
    q = []
    heapq.heappush(q, (0, 0, 1))  # time , cost , here

    while q:
        time, cost, here = heapq.heappop(q)

        if dist[here][cost] < time:
            continue

        for nx, nxc, nxt in adjs[here]:
            nextCost = cost + nxc
            nextTime = time + nxt
            if nextCost > M:
                continue

            if dist[nx][nextCost] > nextTime:
                for i in range(nextCost, M+1):
                    if dist[nx][i] > nextTime:
                        dist[nx][i] = nextTime
                heapq.heappush(q, (nextTime, nextCost, nx))

    return dist[N][M]

T = int(read())
for _ in range(T):
    N, M, K = map(int, read().split())
    adjs = [[] for _ in range(N+1)]
    for _ in range(K):
        x, y, c, b = map(int, read().split())
        adjs[x].append((y, c, b))

    dp = [[INF] * (M+1) for _ in range(N+1)]
    dp[1][0] = 0 # vertex , budget , cost

    for budget in range(M+1) :
        for here in range(1,N+1) :
            if dp[here][budget] == INF : continue

            cost = dp[here][budget]
            for nx,nxc,nxb in adjs[here] :
                if nxc + budget > M : continue
                dp[nx][nxc+budget] = min(dp[nx][nxc+budget],cost+nxb)
        
    ans = min(dp[N])

    if ans == INF :
        print("Poor KCM")
    else :
        print(ans)
